Add Flyout (Popout) Menu to Sidebar + Dropdown on Mobile

krowvin opened the pull request:
This PR aims to provide 2 levels of dropdown (A third with warning). 

It also adds recursion to the traditional sidebar. 

Sidebar docs updated to include these changes. 

If you provide

```jsx
<Sidebar 
...previousProps
enablePopout={true}
popoutDirection="right"
/>
```

![image](https://github.com/user-attachments/assets/2b1bc000-5677-4141-bd7a-dc47cc851cdc)

Leaving `enablePopout={true}` out defaults to false and results in the original with nesting added:
![image](https://github.com/user-attachments/assets/53bd63d8-54a2-4a8a-8967-9f55cfb500af)

![image](https://github.com/user-attachments/assets/aed88663-15c5-42f8-a842-2b081d890903)


- #94 Allow setting ID to target for custom JS/CSS
- #103 
- #89 Added two - three levels deep and blocked beyond this

krowvin commented:
There are a few issues I'm not sure could be considered blocking on this

1. If you click another link right after you click one they both have the arrow int he wrong direction (no reset)
![image](https://github.com/user-attachments/assets/0ab22e7c-3ab4-406f-99bb-9714a56ef624)

2. Pressing Escape closes the popup but does not reset the arrow state until you click anywhere in the page again
![image](https://github.com/user-attachments/assets/1a5a7680-77e3-4ef2-92f8-82cde626660b)

3. The largest issue is how the modal is setup. If it is too far to the right of the screen and the `popoutDirection` is set to "right", it will go outside the viewport. 
![image](https://github.com/user-attachments/assets/f5d22fac-f586-496e-9e15-2ba3323dfdf4)

4. Hot reloading doesn't always play nice with the modal. Because I opted to try and make it backwards compatible I believe this state does not always reset for HMR. 
![image](https://github.com/user-attachments/assets/50667636-fa2c-488d-b439-c35cfd14d3ca)

Probably because i'm using a ref. Might consider removing the ref?

willbreitkreutz commented:
@krowvin easy fix for 1 and 2 above is to not flip the arrow in the first place...  I think that's how i would handle it anyway.

willbreitkreutz commented:
Is there a way to indent the child links in mobile select?  I think you could just do two non-breaking spaces before the link i.e. `  ${text}`

willbreitkreutz commented:
Let's pull out the border-bottom setting on the list items in the sidebar and popout...

FHanbali commented:
> Is there a way to indent the child links in mobile select? I think you could just do two non-breaking spaces before the link i.e. `{backtick ${text}/}backtick`

use ``` ``` to insert in code block

``` ${text} ```

willbreitkreutz commented:
On the swap out of the grid class, let's try to use `md:grid...` rather than add and remove class on a DOM ref.

krowvin commented:
> On the swap out of the grid class, let's try to use md:grid... rather than add and remove class on a DOM ref.  

Updated docs to include this, removed DOM manipulation for this. Tested on current sidebar and sidebar docs page. 

> Let's pull out the border-bottom setting on the list items in the sidebar and popout...  

Removed extra border, ensured border that matches others is present on final dropdown

> easy fix for 1 and 2 above is to not flip the arrow in the first place... I think that's how i would handle it anyway.  

Removed arrow toggle when selected

krowvin commented:
@willbreitkreutz ready for your review

krowvin commented on lib/components/buttons/popover.jsx:
```diff
@@ -79,20 +53,12 @@ function PopoutMenu({ title, children, direction = "right", className }) {
 
       <PopoverPanel
         transition="true"
-        className={`gw-absolute gw-ms-2 gw-pt-1 ${directionClasses[direction]} gw-z-10 gw-mt-2 gw-w-56 gw-shrink gw-rounded-xl gw-bg-white gw-text-sm gw-leading-6 gw-text-gray-900 gw-shadow-lg gw-ring-1 gw-ring-gray-900/5`}
```
I also fixed the padding, they were spaced too far apart imo

willbreitkreutz commented on lib/components/form/dropdown.jsx:
Is this meant to be exported as part of the form group of components?

krowvin replied:
No, 

Will make sep PR to create docs and make it it's own thing

krowvin replied:
Future work here
#113

willbreitkreutz commented on lib/components/layout/usace-box.jsx:
I'm not clear on the use-case here, but likely this is a candidate for forwardRef https://react.dev/reference/react/forwardRef

krowvin replied:
Converted to forwardRef, was not aware this existed but I see why now! 

Needed a ref for usacebox so that it could persist between renders. Sidebar was setup to include the usace box:
https://github.com/USACE/groundwork/blob/5a429e84f7d0846180c74285dc89404f177bfe8c/lib/composite/sidebar/index.jsx#L135

 between page re-render (mobile vs not / i.e. page resize)

willbreitkreutz commented on lib/composite/sidebar/index.jsx:
```diff
@@ -1,51 +1,180 @@
+            className={"gw-w-5/6 gw-m-auto"}
+            value={selectedPath}
+            onChange={(e) => {
+              doUpdateHash(e.target.value);
```
I think this breaks us if we're not in the GH-pages env and aren't using hash based navigation

willbreitkreutz replied:
might have to have a top-level prop for a nav function or something if we don't want to use an a tag

willbreitkreutz reviewed (commented):
Some stuff to chat about, but overall good

krowvin commented:
@willbreitkreutz 

* Removed <BlackDot /> that slipped in on the header branch (Removed it from this one and forgot to rebase the other branch)
* Fixed merge conflicts with main. 
* Removed onClick in favor of a-tag
* Removed doUpdateHash with a-tag click
* Updated sidebar docs
* Tested mobile/desktop functionality
